operator/internal/validation/lokistack.go

```diff
@@ -115,6 +120,89 @@ func (v *LokiStackValidator) validateHashRingSpec(s lokiv1.LokiStackSpec) field.
 	return nil
 }
 
+func (v *LokiStackValidator) validateLimitsSpec(s *lokiv1.LimitsSpec) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	if s.Global == nil && s.Tenants == nil {
+		return nil
+	}
+
+	var errors field.ErrorList
+	if s.Global != nil && s.Global.IngestionLimits != nil {
+		fp := field.NewPath("spec", "limits", "global", "ingestion", "desiredRate")
+		errors = append(errors, validateIngestionLimits(s.Global.IngestionLimits, fp)...)
+	}
+
+	for id, tenant := range s.Tenants {
+		if tenant.IngestionLimits == nil {
+			continue
+		}
+
+		fp := field.NewPath("spec", "limits", "tenants", id, "ingestion", "desiredRate")
+
+		if s.Global == nil {
+			errors = append(errors, validateIngestionLimits(tenant.IngestionLimits, fp)...)
+		} else {
+			errors = append(errors, validateTenantIngestionLimits(s.Global.IngestionLimits, tenant.IngestionLimits, id, fp)...)
+		}
+	}
+
+	return errors
+}
+
+// validateIngestionLimits checks if a given IngestionLimitSpec is valid for:
+// - Fields `desiredRate` and `perStreamRateLimit` not both set at a time.
+func validateIngestionLimits(spec *lokiv1.IngestionLimitSpec, invalidFieldPath *field.Path) (errs field.ErrorList) {
+	// Check against per stream rate limit
+	if spec.DesiredRate > 0 && spec.PerStreamRateLimit > 0 {
+		errs = append(errs, field.Invalid(
+			invalidFieldPath,
+			spec.DesiredRate,
+			lokiv1.ErrDesiredRateAndPerStreamRateNotAllowed.Error(),
+		))
+	}
+
+	return errs
+}
+
+// validateTenantIngestionLimits checks if a given tenant IngestionLimitSpec is valid for:
+// - Tenant fields `desiredRate` and `perStreamRateLimit` not both set at a time.
+// - Global field `desiredRate` and tenant field `perStreamRateLimit` not both set at a time.
+// - Tenant field `desiredRate` and global field `perStreamRateLimit` not both set at a time.
```

Comment:
Am I understanding this correctly that the configuration must be "either all use desiredRate or all use perStreamRateLimit" and there's never a valid combination which uses desiredRate on one side and perStreamRateLimit on the other?

Reply:
It's not a choice of valid vs. invalid although the webhook implies this. My call here is to keep things uniform and simple (as we try to keep the stacks similar), either you opt-in for Loki doing the sharding or you keep managing the rates yourself.